### ISSUE TITLE
Internal Registry Authentication Issues

### ISSUE THREAD

**Issue by StevenBarre** [BCDevOps/OpenShift4-RollOut#428]
**Describe the issue**
Because we've added an [externally accessible route](https://app.zenhub.com/workspaces/platform-experience-5bb7c5ab4b5806bc2beb9d15/issues/bcdevops/openshift4-rollout/309) to the Image Registry it is causing some issues with pulling images.

The default service accounts `.dockercfg` secret only contain tokens for the internal name `image-registry.openshift-image-registry.svc:5000` while the `oc get is` command returns the external name. So a user just using copy/paste will be unable to pull images from their own namespace, let alone any others that they have access to.

**Which Sprint Goal is this issue related to?**
SG - Sprint 19 Platform Hot Fixes

**Additional context**
https://github.com/BCDevOps/OpenShift4-Migration/issues/14

**Definition of done Checklist (where applicable)**
- [ ] Make life easy for developers

**Comment by StevenBarre** [comment]
Opened case 02787852 with Red Hat. Blocked waiting on a response.

**Comment by jefkel** [comment]
Question:
Do we want to enable this path for developers?  (it's not an optimal solution or approach)

Current recommended practice:
Manifests deployed on a cluster use the internal service name for the registry
- Authentication is automatic for cluster service accounts
- reduces traffic, and dependency on cluster default ingress.


Developer teams to use the right names for the appropriate use
- track the OCP4 internal cluster name and use that - this service name will be the same on all OCP4 clusters.
- For out-of-cluster access, the external name/explicit user/token is used - Service names will be different for all OCP clusters.

**Comment by StevenBarre** [comment]
> Do we want to enable this path for developers?

I think we want devs to be able to just copy/paste from `oc get is`. Having that fail is non-intuitive. Maybe having `oc get is` return the internal repo would be better?

Or maybe just having the internal repo in the default search path so you can just put `namespace/image` without the hosnmame?

**Comment by jefkel** [comment]
Is there a reason why we don't have the internal repo first in the default search path?  (feels like this could be a decent developer experience improvement)

If we make the easy-path work (default search path start with internal registry), that will ideally help the developers use that approach first, and it gives us a decent documentation point for when people cut and paste the output from the oc cli.  (ie: we can educate them that the oc results are for out of cluster, for default authentication inside the cluster, simply remove the DNS and keep the path/image:tag format.)

**Comment by StevenBarre** [comment]
> Is there a reason why we don't have the internal repo first in the default search path?

Because OCP didn't come configured like that out of the box. I've opened a RH case to ask if there is any reason we can't add it ourselves with a MachineConfig https://access.redhat.com/solutions/5063741

**Comment by StevenBarre** [comment]
> Is there a reason why we don't have the internal repo first in the default search path?

Matt says the reason why it's not included is that it is bad practice to use image paths without a specific registry included.

**Comment by StevenBarre** [comment]
Still waiting on a response to the first case. Matt is going to take over from the other support person.

**Comment by StevenBarre** [comment]
Our best path here may be to just get rid of the external route once we get Artifactory running. Then `oc get is` will return the correct URL and everything will work.

**Comment by StevenBarre** [comment]
From RH:

>Discussing with Engineering, the secret is not created intentionally because we do not want internal components (kubelet) to use the router router to pull images from the internal registry as it's reduces reliability of the registry.
>
>Pods should use the Service, i.e. image-registry.openshift-image-registry.svc:5000 even if external routes are configured.

We've added docs to advise devs about this in https://github.com/BCDevOps/OpenShift4-Migration/issues/14

I don't think we should be adding the internal registry to the default search path, as that enables bad practice.

@jefkel OK with you if we close this WONTFIX ?

**Comment by jefkel** [comment]
I'm happy with it.. (as long as we keep that distinction and outcome super clear in the docs - ie: deployment configurations should NOT use the external registry URL for the internal registry service)
